// @rc-component/util/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "get", {
  enumerable: true,
  get: function () {
    return _get.default;
  }
});
Object.defineProperty(exports, "set", {
  enumerable: true,
  get: function () {
    return _set.default;
  }
});
Object.defineProperty(exports, "supportNodeRef", {
  enumerable: true,
  get: function () {
    return _ref.supportNodeRef;
  }
});
Object.defineProperty(exports, "supportRef", {
  enumerable: true,
  get: function () {
    return _ref.supportRef;
  }
});
Object.defineProperty(exports, "useComposeRef", {
  enumerable: true,
  get: function () {
    return _ref.useComposeRef;
  }
});
Object.defineProperty(exports, "useEvent", {
  enumerable: true,
  get: function () {
    return _useEvent.default;
  }
});
Object.defineProperty(exports, "useMergedState", {
  enumerable: true,
  get: function () {
    return _useMergedState.default;
  }
});
Object.defineProperty(exports, "warning", {
  enumerable: true,
  get: function () {
    return _warning.default;
  }
});
var _useEvent = _interopRequireDefault(require("./hooks/useEvent"));
var _useMergedState = _interopRequireDefault(require("./hooks/useMergedState"));
var _ref = require("./ref");
var _get = _interopRequireDefault(require("./utils/get"));
var _set = _interopRequireDefault(require("./utils/set"));
var _warning = _interopRequireDefault(require("./warning"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
